packages/ckeditor5-engine/src/conversion/downcasthelpers.js

```diff
@@ -755,6 +863,141 @@ function removeUIElement() {
 	};
 }
 
+// Function factory that creates a default converter for model markers.
+//
+// See {@link DowncastHelpers#markerToData} for more information what type of view is generated.
+//
+// This converter binds created UI elements and affected view elements with the marker name
+// using {@link module:engine/conversion/mapper~Mapper#bindElementToMarker}.
+//
+// @returns {Function} Add marker converter.
+function insertMarkerData( viewCreator ) {
+	return ( evt, data, conversionApi ) => {
+		const viewData = viewCreator( data.markerName );
+
+		if ( !viewData ) {
+			return;
+		}
+
+		const markerRange = data.markerRange;
+
+		if ( !conversionApi.consumable.consume( markerRange, evt.name ) ) {
+			return;
+		}
+
+		const mapper = conversionApi.mapper;
+		const viewWriter = conversionApi.writer;
+
+		// Adding closing data first to keep the proper order in the view.
+		handleMarkerBoundary( markerRange, false );
+		handleMarkerBoundary( markerRange, true );
+
+		evt.stop();
+
+		function handleMarkerBoundary( range, isStart ) {
```

Comment:
I don't like that this function is defined here. It'll be create on every call of `insertMarkerData`. Not that I see performance issues already, but just for the sanity I'd move it outside (if possible).

Reply:
Same with the other helpers.

Reply:
I had the same feelings as you but for clarity reasons, I left those helpers inside those functions. They are used only by those functions and this way they can receive fewer parameters.